[MRG + 1] implementing LDA(Latent Dirichlet Allocation) with online variational Bayes

chyikwei opened the pull request:
This PR is an implementation of Matt Hoffman's topic modeling algorithm `LDA with online variational Bayes`.

Based on previous discussion in [this email thread](http://sourceforge.net/p/scikit-learn/mailman/message/32748341/), I asked Matt if he could relicense his `onlineldavb` code to BSD. And now his code is relicensed, so I create a PR for it.

I use the name `OnlineLDA` for this model and put it in `decomposition` folder. And since the model can run both online and batch update, I implemented both `fit` and `partial_fit` method.
The algorithm part and unit test is done and ready for review. Will work on an example next.
###### Check List:
- [X] algorithm implementation
- [X] unit test 
- [X] profiling
- [x] optimization
- [X] example
- [x] documentation
###### Reference:

[1] ["Online Learning for Latent Dirichlet Allocation", Matthew D. Hoffman, David M. Blei, Francis Bach](https://www.cs.princeton.edu/~blei/papers/HoffmanBleiBach2010b.pdf)
[2] [original onlineldavb code (with BSD license)](https://www.dropbox.com/s/wnkro3xtqjm7bli/onlineldavb_bsd.tar)

chyikwei committed:
add OnlineLDA model and test

chyikwei committed:
don't use sparse.block_diag since it is new in scipy 0.11.0

update params and minor change

add onlineLDA example

rename onlineLDA to LatentDirichletAllocation

raise NotFittedError in trnasform

change model name in LDA example

cython optimization for online LDA model

fix import error

add evaluate_every parameter in online LDA model

add score in LDA model

fix doc format

update variable name and improve docs

fix broken example and change parameter name to n_samples & n_features

add check_non_negative function in LDA

add parameter to specify online/batch VB in fit method

improve documentation

set default hyper parameter to 1 / n_topics

use gen_batches in example

remove old func from example

amueller commented:
Can we please rename this to ``LatentDirichletAllocation`` even though that is long? We have a (badly named) ``LDA`` class in scikit-learn.

amueller commented:
Not sure decomposition is the right folder, but I don't have a better idea ^^

amueller commented:
How does this compare against the gensim implementation? Is that the same approach?

chyikwei commented:
@amueller 
1) ok. I will rename it.
2) I am not sure about the folder either, but "decomposition" is the best one I can find.
3) yes. based on [gensim's LDA page](https://radimrehurek.com/gensim/models/ldamodel.html), I think its approach is also M. Hoffman's online LDA.

amueller commented:
For 3) it would be cool if you could give a performance comparison (and maybe also a comparison of how well it fit the data?) as a sanity check?

chyikwei commented:
ok. will add performance comparison with gensim's implementation. For "how well it fit the data", I will compare perplexity.

amueller commented:
Thanks :)

ariddell commented:
I've almost got the transform method working for LDA in https://github.com/ariddell/lda ; I would imagine Gibbs sampling beats online LDA in perplexity and is competitive with NMF in speed for small to medium datasets -- and I'd be very curious to see how things play out with large datasets.

chyikwei commented:
quick update:
1) renamed model to LatentDirichletAllocation
2) add some cython optimization
3) will work on performance comparison with gensim next. (not familiar with its interface yet.)

@ariddell thx! I will start with 20 news group data set first, and we can try large one later.

amueller commented:
It would be very interesting to see how the collapsed gibbs sampler compares to this, indeed.

ariddell commented:
For larger datasets, there's Enron and PubMed: http://archive.ics.uci.edu/ml/datasets/Bag+of+Words

chyikwei commented:
Hi,
I put the comparison between my LDA implementation and gensim in [this spreadsheet](https://docs.google.com/spreadsheets/d/1tkN-j7aePRsWK-EEBHjTNTdTmYCFSgN2rzVj09LLqb8/edit?usp=sharing).  (My script is [here](https://gist.github.com/chyikwei/34b97d4d443a0cc38a2f).)

I use 20 newsgroup dataset and compared both online and batch update.
In batch mode, the performance is close.
In online mode, the speed difference is caused by how often we compute perplexity in the training step.
My `partial_fit` method doesn't compute perplexity at all, so it is much faster than gensim. (For detail, check notes in the result spreadsheet.)

Note: One thing I haven't figured out is why perplexity goes up in gensim as the number of workers increased. I will double check that

@ariddell It will be cool if you can add Gibbs sampler's result. I will check the larger datasets link you post next.

GaelVaroquaux commented:
Is there a reason that perplexity is computed by default at every step?
It seems to me that we could make model convergence faster by computing
it every 2 or 4 step.

GaelVaroquaux commented:
Well, the spreadsheet is overall in favor of your implementation. Good work!

amueller commented:
yeah that looks promising :)
It might be nice to have a ``plot_`` example that generates an image. Then the output will be visible in the example gallery on the website.

chyikwei commented:
@GaelVaroquaux There is no reason to compute perplexity in every step. I will add a parameter for this (similar to gensim's `eval_every `). Also, I think I need to check gensim's source code and setting to make sure I am doing apple to apple comparison. (If we both use Matt Hoffman's code, the result should be similar.)

@amueller not sure what's the best way to visualize topic models. (usually, I just check top words in each topic.) Any idea?

amueller commented:
Maybe pick the top three words in each topic and then to a bar-graph on how likely they are under each of the topics?

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,522 @@
+
+        return score
+
+    def preplexity(self, X, gamma, sub_sampling=False):
```
It would make sense to have a ``score`` method based on ``transform`` and ``perplexity``, right?

chyikwei replied:
yes. that make sense. will add it.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,522 @@
+
+    Parameters
+    ----------
+    n_topics: int, optional (default: 10)
```
There needs to be a space before ``:`` for the numpy doc format (yes, everywhere, sorry).

chyikwei replied:
k. will fix it

amueller commented:
This might be a stupid question, but if we wanted to add the collapsed gibbs sampler version, say the one by @ariddell, could we use the same public interface and branch using an ``algorithm='collabsed_gibbs'`` parameter?

chyikwei commented:
yes. I think we can share interface for different implementation. `online variational Bayes` algorithm just have a few more parameters used for online update, which can be ignored by `gibbs sampler`.

amueller commented:
Ok cool :)

chyikwei commented:
quick update:

1. added `evaluate_every ` parameter to control how often we calculate perplexity.<br>
In my test, training is about 2x faster when I disable evaluation in my test. (58s vs 101s)
2. add `score` method and fix document format.
3. updated performance comparison spreadsheet.<br>
I found `gensim` use `chunksize * workers` for each update ([link](https://github.com/piskvorky/gensim/blob/develop/gensim/models/ldamulticore.py#L178)), so the performance get worse as `workers` increases. The updated result is [here](https://docs.google.com/spreadsheets/d/1tkN-j7aePRsWK-EEBHjTNTdTmYCFSgN2rzVj09LLqb8/edit?usp=sharing).(check blue rows, the online performance for different worker are similar now.)

ariddell commented:
This is really great. Two comments

1. `s/preplexity/perplexity/g`
2. Should the behavior be consistent with NMF? `transform` in that case gives back the "document-topic" decomposition [n_docs, n_topics]. Is that what is happening here?

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,530 @@
+
+        Returns
+        -------
+        gamma : array, [n_dics, n_topics]
```
n_docs

amueller commented:
I think this looks great, but one thing I just noticed is the use of greek letters as parameter and variable names. I know that is something that is quite common and if you are part of the community, you know what it means. I think it would be better to give telling names, like ``topic_prior`` instead of ``eta`` or something like that.

chyikwei commented:
@ariddell @amueller 
Thanks for review the code and find those errors!
Will fix misspelled word and typo and change parameter/variable names.

ariddell commented:
I think `eta` and `alpha` are standard by this point. It would be great to have other names, but I haven't seen a good replacement. There are actually two priors and both relate to topics (one is the topic-word Dirichlet and the other is document-topic Dirichlet)

amueller commented:
Why not call them ``topic_word_prior`` and ``document_topic_prior`` then?

Someone who is familiar with the literature will immediately know what they are, and someone not familiar with the literature will at least have some idea. And the doc-strings can mention the standard notation.
Or do you think it is very confusing for practitioners?

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,530 @@
+        set it to -1 to not evalute preplexity in training at all.
+
+    normalize_doc : boolean, optional (default: False)
+        normalize the topic distribution for transformed document or not.
```
Capitalize please.

amueller commented:
I'll try do do an actual review pretty soon, catching up with github issues from the holidays.

ariddell commented:
I suppose `topic_word_prior` and `document_topic_prior` are ok. Some of the greek names are familiar and some of them are not. `alpha` and `eta` happen to be standard. I suspect that if you're not going to use the defaults, you'll be familiar with the greek letters.

chyikwei commented:
No sure why the test failed in `FAIL: sklearn.tests.test_common.test_classifiers('LinearSVC', <class 'sklearn.svm.classes.LinearSVC'>)` (Python 3.4 environment). 
Will need to setup a Python 3.4 environment and check.

chyikwei commented:
Looks like my previous test failed because `random_state` is not set in `check_classifiers_train` test.
It passed after I add `set_random_state(classifier)`.
Not sure if I should open a separate issue/pull_request for that.

amueller commented:
I just pushed a fix for that to master as 45982e0. You can rebase if you like.

chyikwei commented:
rebased. thx

larsmans commented:
+1 for `topic_word_prior` and `document_topic_prior`.

On another note, this should be using `n_samples` instead of `n_docs` and `n_features` instead of `n_vocabs`. There's no such thing as a "vocab", that's just an abbreviation of "vocabulary" that we sometimes use. What you mean is either "features" or "words".

larsmans commented:
The example is broken: `unexpected keyword argument 'tau'`

larsmans commented:
More generally, I would love to see a high-performance LDA in scikit-learn; I'm now using Gensim, which takes half an hour just to establish the vocabulary of a set of 6e5 documents that `CountVectorizer` parses in just 15s.

chyikwei commented:
@larsmans thx. I forgot to change the example last time. fixing now.will also change parameter name to `n_samples` and `n_features`.

flaviovdf commented:
I just checking out this code (usually I lurk around sklearn mailing list). I think this is great, since I constantly use LDA (mostly on gensim), but prefer sklearn's api. One minor suggestion I have is to check that the input are non-negative integers only. LDA only make's sense with this input. I just checked that the NMF code does something similar.

chyikwei commented:
@flaviovdf yes. that make sense. will add non-negative check. thx

scharron commented:
I tried this LDA implementation on a dataset of 10010 documents / 289712 words.
I played the test with 10 topics (not enough to be of quality)
Results: https://gist.github.com/scharron/22ce0ebbced8b60665d8
@ariddell's implementation is too slow with this number of documents (mainly when initializing) (with ~3000 documents, it took 3 hours IIRC)

I also tried this implementation with a more realistic case of ~100 topics. It's still too slow for us (about 2 hours IIRC) to use it right now (and we would like to play that on 100000 / 1000000 documents, with about the same vocabulary size)

ariddell commented:
Yeah, Gibbs is slower. VB is a better fit for Scikit-learn. The model is going to be called OnlineLDA, right? 

The shortcomings of VB vis-a-vis MCMC are well-known (or at least google-able). Briefly, you miss a lot of the dependencies between parameters.

chyikwei commented:
@scharron did you try `partial_fit` method?

In my current implementation, `fit` method use `batch VB` while `partial_fit` method use `online VB`. `online VB` will converge faster than `batch VB` when the dataset is large. You can check the experiment result from the original online LDA paper [here](https://www.cs.princeton.edu/~blei/papers/HoffmanBleiBach2010b.pdf).

mblondel commented:
@scharron You might be interested in trying my NMF implementation:
https://gist.github.com/mblondel/09648344984565f9477a

IIRC, when I tried it on the News20 dataset, it finished under 20 seconds.

NMF works quite nicely for topic modeling.

chyikwei commented:
@scharron Also, if you use `fit` method, set `evaluate_every` parameter to `-1` or a larger number will save some time.

scharron commented:
@mblondel It's fast indeed! (~1 minute on my dataset, 100 topics, it stopped after 13 iterations). It seems of a lower quality (at least for the most significant terms of each topic), but good enough to be of interest :-)
Thanks !

scharron commented:
@chyikwei ok it's better with online LDA (70 seconds for the 10-topics example train set, 251 seconds for 100 topics)

When I did my previous test with 100 topics (2 hours), the evaluate_every was at 5.
I'll read the paper, thanks !

amueller commented:
Side-note: This model will not be called OnlineLDA but ``OnlineLatentDirichletAllocation`` (I wouldn't mind just ``LatentDirichletAllocation``). We are just renaming ``sklearn.lda.LDA`` to ``sklearn.discriminant_analysis.LinearDiscriminantAnalysis`` ;)

amueller commented:
Sorry, what was the status of the PR? I kinda lost track a bit :-/ needs reviews?

chyikwei commented:
@amueller for now I use the name `LatentDirichletAllocation` since it includes both batch and online VB update.
And for status, I think it is waiting for reviews now.

jnothman commented:
It sounds like `evaluate_every=5` is not a reasonable default. I also think the docstring for that param needs to explain why evaluating perplexity may beneficial.

To what extent does that account for the huge batch training time relative to online? Would it be useful to have a parameter that specifies online training in batches should be used when `fit` is called, such that fast online training can be used in a pipeline (where `partial_fit` is not currently supported)?

amueller commented:
@chyikwei sorry, there is a huge backlog of high-quality contributions, but this one is very high on my list to review.

mblondel commented:
@scharron This is probably because my implementation uses the squared loss. I'll try to add the generalized KL divergence when I get time. It's known to get better topics in general.

chyikwei commented:
@jnothman 
1) For `evaluate_every` parameter:
I agree that `5` is not a good default value.I think a reasonable value should also depend on `max_iters` in the training step. 
Maybe I can change it to `evaluate_frequency` and let `evaluate_every = max_iters / evaluate_frequency`. For example, when `max_iters=10` and `evaluate_frequency=2.0`, we will get `evaluate_every=(10/2.0)=5`.

2) For batch vs online method:
In batch method, huge training time comes from slow convergence, multiple iterations, and perplexity evaluation. (online method only run 1 iteration and doesn't evaluate perplexity.)
Add yes, `fit` function should converge faster with online update. I will add a parameter for that.

@amueller 
There's no rush. I know there are many issues and PRs in this project.

amueller commented:
For 2) I would suggest adding "batch_size=1000" or something, and when it is None, the whole dataset is used.

amueller commented on examples/decomposition/topics_extraction_with_online_lda.py:
```diff
@@ -0,0 +1,77 @@
+    Example for LDA online update
+    """
+
+    def chunks(l, n):
```
maybe use the gen_batches method?

amueller replied:
You can remove the function now, I think.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,653 @@
+    """
+    Online Latent Dirichlet Allocation implementation with variational inference
+
+    References
```
This should be below the parameters

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,653 @@
+
+    learning_offset : float, optional (default: 1000.)
+        A (positive) parameter that downweights early iterations in online learning.
+        It should be greater than 1.0. In literature, it is parameter `tau0`.        
```
I think the formulation should be "In the literature, this is called tau0" [also in other places]

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,653 @@
+
+    Parameters
+    ----------
+    n_topics : int, optional (default: 10)
```
I think we usually use (default=10), so ``=`` not ``:  ``

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,653 @@
+    n_topics : int, optional (default: 10)
+        Number of topics.
+
+    doc_topic_prior : float, optional (defalut: 0.1)
```
default

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,653 @@
+        Number of topics.
+
+    doc_topic_prior : float, optional (defalut: 0.1)
+        Prior of document topic distribution `theta`. In general, it is `1 / n_topics`.
```
I'm not sure if "in general" is a good formulation. Shouldn't this default be coupled to the default of ``n_topics``, so that is someone changes ``n_topics`` the ``doc_topic_prior`` also changes?

chyikwei replied:
Yeah. I will set default to `1 / n_topics` if `doc_topic_prior` is not specified.
And Looks like [gensim](https://github.com/piskvorky/gensim/blob/develop/gensim/models/ldamodel.py#L198-L200) set default to `1 /n_topics` too.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,653 @@
+    learning_decay : float, optional (default: 0.7)
+        It is a parameter that control learning weight for `_component` in online learning.
+        The value should be set between (0.5, 1.0] to guarantee asymptotic convergence.
+        When the value is 0.0 and batch_size is `n_samples`, the udpate is same as batch learning.
```
"update"

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,653 @@
+        The maximum number of iterations.
+
+    total_samples : int, optional (default: 1e6)
+        Total umber of document. It is only used in `partial_fit` method.
```
number of documents

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,653 @@
+        Total umber of document. It is only used in `partial_fit` method.
+
+    batch_size : int, optional (default: 128)
+        Number of document to use in each EM iteration. Only used in online learning method.
```
documents

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,653 @@
+        Number of document to use in each EM iteration. Only used in online learning method.
+
+    evaluate_every : int optional (default: 5)
+        How many iterations to evaluate perplexity once. Only used in `fit` method yet.
```
"How often to evaluate perplexity"

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,653 @@
+    batch_size : int, optional (default: 128)
+        Number of document to use in each EM iteration. Only used in online learning method.
+
+    evaluate_every : int optional (default: 5)
```
I think the default should be -1

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,653 @@
+        Evaluating perplexity in every iteration will increase training time up to 2X.
+
+    normalize_doc : boolean, optional (default: False)
+        Normalize the topic distribution for each document or not.
```
Whether to normalize the topic distribution for each document.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,653 @@
+        Max number of iterations for updating document topic distribution in E-step.
+
+    n_jobs : int, optional (default: 1)
+        Number of parallel jobs to run in E-step. -1 for autodetect.
```
-1 for using all cores, I think (or copy whatever docstring there is usually)

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,653 @@
+
+        Returns
+        -------
+        doc_topic_distr : array, [n_samples, n_topics]
```
``shape=(n_samples, n_topics)`` is the preferred form, I think.

amueller commented on sklearn/decomposition/tests/test_online_lda.py:
```diff
@@ -0,0 +1,294 @@
+
+def _build_sparse_mtx():
+    """
+    Create 3 topics and each have 3 words
```
please use comments and not docstrings in tests, so
# Create 3 topics ..

instead of
``"""``.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,653 @@
+            Topic distribution for each document.
+        """
+        X = self._to_csr(X, "LatentDirichletAllocation.fit_transform")
+        return self.fit(X).transform(X)
```
this could be implemented more efficiently by skipping the additional E-step, right?
Or is that instantaneous anyhow?

You could rename ``fit`` to ``_fit``, return ``doc_topics`` in ``_fit``, here just return ``_fit`` and in ``fit`` just call ``_fit`` and then ``return self``.

amueller replied:
Similar to ``PCA`` but I think we had another place were that was done.

amueller commented:
Thanks for the fixes. I'll do a proper review soon :)

amueller commented:
Only sort-of related to the PR, does anyone know if there is an online collapsed VB version?

chyikwei commented:
@amueller Not sure if there is online CVB version, but [this paper](http://arxiv.org/pdf/1206.6425v1.pdf) might be related. (It use online VB and Gibbs sampling together to achieve sparse  topic-word update.)

amueller commented on examples/decomposition/topics_extraction_with_online_lda.py:
```diff
@@ -0,0 +1,71 @@
+Topics extraction with Latent Dirichlet Allocation
+==================================================
+
+This is an example of showing how to extrat topics from
```
extract

amueller commented on examples/decomposition/topics_extraction_with_online_lda.py:
```diff
@@ -0,0 +1,71 @@
+20 newsgroup dataset with OnlineLDA. Also, It shows how 
+top words in each topic changed during training.
+
+The code is modified from
```
from the

amueller commented:
could you please rebase on master? That was a fix in the dataset loading.

amueller commented on examples/decomposition/topics_extraction_with_online_lda.py:
```diff
@@ -0,0 +1,71 @@
+    Example for LDA online update
+    """
+
+    # In default, we set topic number to 10, and both hyperparameter
```
"By default, we set the number of topics to 10". If it is by default, why do we set it here?

amueller commented on examples/decomposition/topics_extraction_with_online_lda.py:
```diff
@@ -0,0 +1,71 @@
+    # eta and alpha to 0.1 (`1 / n_topics`)
+    n_topics = 10
+
+    # chunk_size is how many records we want to use
```
maybe documents instead of records?

amueller commented on examples/decomposition/topics_extraction_with_online_lda.py:
```diff
@@ -0,0 +1,71 @@
+            doc_mtx = vectorizer.transform(doc_list)
+
+        # fit model
+        print("\nFitting LDA models with online udpate on chunk %d..." %
```
update

amueller commented on examples/decomposition/topics_extraction_with_online_lda.py:
```diff
@@ -0,0 +1,71 @@
+
+
+if __name__ == '__main__':
+        online_lda_example()
```
indented too far, should be 4 chars

amueller commented on sklearn/decomposition/_online_lda.pyx:
```diff
@@ -0,0 +1,73 @@
+from scipy.special import psi
+
+
+@cython.boundscheck(False)
```
maybe move them at the top of the file?

amueller commented on sklearn/decomposition/_online_lda.pyx:
```diff
@@ -0,0 +1,73 @@
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def mean_change(np.ndarray[ndim=1, dtype=np.float64_t] arr_1,
```
I guess you benchmarked it, what is the speedup of doing this vs ``np.abs(arr_1 - arr_2).mean()``? I guess you save the allocation of two temporary arrays? Not sure since when ``abs`` has an ``out`` and if that would make it better....

amueller commented on sklearn/decomposition/_online_lda.pyx:
```diff
@@ -0,0 +1,73 @@
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def _dirichlet_expectation_1d(np.ndarray[ndim=1, dtype=np.float64_t] arr):
```
Again, did you benchmark?
This is ``psi(arr) - psi(arr.sum())``, right?
Can you please add this as documentation?

amueller commented on sklearn/decomposition/_online_lda.pyx:
```diff
@@ -0,0 +1,73 @@
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def _dirichlet_expectation_2d(np.ndarray[ndim=2, dtype=np.float64_t] arr):
```
Again, it would be nice to know the benefit, and add a docstring that explains the computation,
``psi(arr) - psi(arr.sum(axis=))``, I think.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+                          _dirichlet_expectation_2d)
+
+
+def check_non_negative(X, whom):
```
If this is a copy from NMF, it should probably be moved to ``sklearn.utils.validation`` and imported in both places.

larsmans replied:
Either that, or the name should be `_check_non_negative`. (NMF gets this wrong, but that's another issue.)

amueller replied:
Or both. make it private and move to utils

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+        raise ValueError("Negative values in data passed to %s" % whom)
+
+
+def _dirichlet_expectation(X):
```
maybe ``_log_dirichlet_expectation``?

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+    Parameters
+    ----------
+    X : array-like
+        1 or 2 dimention vector
```
dimensional vector

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+
+    Returns
+    -------
+    dirichlet_expect: array-like
```
for consistency maybe a space before ":"

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+    X : sparse matrix, shape=(n_samples, n_features)
+        Document word matrix.
+
+    exp_topic_word_distr : dense matrrix, shape=(n_topics, n_features)
```
matrix

larsmans replied:
array.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+        Stopping tolerance for updating document topic distribution in E-setp.
+
+    cal_diff : boolean
+        Parameter that indicate to calculate differene in `component_` or not.
```
Whether to calculate the change in ``component_``.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+
+    cal_diff : boolean
+        Parameter that indicate to calculate differene in `component_` or not.
+        Set `cal_diff` to `True` when we need to run M-step.
```
I'm not sure if I understand this comment. The M-step needs ``component_diff``?

chyikwei replied:
I think the variable name `cal_diff` is misleading.
I will change it to `cal_sstats` since in the original code it is called [sstats](https://github.com/qpleple/online-lda-vb/blob/master/mdhoffma/onlineldavb.py#L189-L191).
And it corresponds to `sum( n_{dw} * phi_{dwk})` in M-step.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+    Returns
+    -------
+    (doc_topic_distr, component_diff) :
+        `doc_topic_distr` is unnormailzed topic distribution for each document.
```
"is the unnormalized"

amueller commented:
I just realized this implementation currently only works for sparse data. That would be a new one for scikit-learn. Maybe it would be worth implementing the dense variant, too, before merging? I realize that this is mostly used for text data and the sparse case is much more interesting, but it might be a bit odd to have a model that doesn't work on dense data.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+
+        # The optimal phi_{dwk} is proportional to 
+        # exp(E[log(theta_{dk})]) * exp(E[log(beta_{dw})]).
+        norm_phi = np.dot(exp_doc_topic_d, exp_topic_word_d) + 1e-100
```
Maybe define a constant EPS for 1e-100? Or use ``np.finfo(np.float).eps``?

amueller replied:
why do you not compute this at the beginning of the loop? You have it once before the loop and once at the end.

amueller replied:
why is this called norm_phi, it is just phi in the paper, right?

amueller replied:
I guess this is related to the comment

                # We represent phi implicitly to save memory and time.
                # Substituting the value of the optimal phi back into
                # the update for gamma gives this update. Cf. Lee&Seung 2001.

I haven't cf. lee & seung. That's the NMF nature paper, right? I don't immediately see the connection.

chyikweiyau replied:
To my understanding, Hoffman just don’t want to store the `phi` vector.
And since  `phi_{dwk} proportional to expElog( theta_{dk}) * expElog( theta_{dk})`,
we can rewrite `sum( phi_{dwk} * n_{dw} )`
to `sum( (expElog( theta_{dk}) * expElog( theta_{dk}) / phi_norm) * n_{dw})`
which leads to the `gamma_{dk}` update in his original code [here]
(https://github.com/qpleple/online-lda-vb/blob/master/mdhoffma/onlineldavb.py#L179-L180).

And yes, the reference is the NMF paper. Maybe he is inspired by the updating rules sections. (Section 4 & 5)

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+        for it in xrange(0, max_iters):
+            last_d = doc_topic_d
+
+            doc_topic_d = doc_topic_prior + exp_doc_topic_d * \
```
Can you please use parentheses around the expression instead of ``\``?

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+    n_topics : int, optional (default=10)
+        Number of topics.
+
+    doc_topic_prior : float, optional (defalut=None)
```
*default

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+        Number of topics.
+
+    doc_topic_prior : float, optional (defalut=None)
+        Prior of document topic distribution `theta`. If the value is None, default
```
defaults

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+        In the literature, this is called `alpha`.
+
+    topic_word_prior : float, optional (default=None)
+        Prior of topic word distribution `beta`. If the value is None, default
```
defaults

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+
+    learning_method : 'batch' | 'online', default='online'
+        Method used to update `_component`. Only used in `fit` method.
+        In general, if the data size is large, online update will be much faster than
```
the online update will be much faster than the batch update.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+        Valid options::
+
+            'batch': Batch variational bayes method. Use all training data in each EM udpate.
+                Old `components_` will be overrided in each iteration.
```
overwritten

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+            'batch': Batch variational bayes method. Use all training data in each EM udpate.
+                Old `components_` will be overrided in each iteration.
+            'online': Online variational bayes method. In each EM update, use min batch of
+                training data to update `components_` variable incrementally. The weight in
```
to update the ``components_`` variable

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+                Old `components_` will be overrided in each iteration.
+            'online': Online variational bayes method. In each EM update, use min batch of
+                training data to update `components_` variable incrementally. The weight in
+                each udpate is controlled by `learning_decay` and `learning_offset` parameter
```
update

amueller replied:
by the ``learning_decay`` and ``learning_offset`` parameters.

amueller replied:
I think it should be learning rate not learning weight. Or step size.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+                each udpate is controlled by `learning_decay` and `learning_offset` parameter
+
+    learning_decay : float, optional (default=0.7)
+        It is a parameter that control learning weight for `_component` in online learning.
```
again, I think learning rate or step size would be better.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+        The maximum number of iterations.
+
+    total_samples : int, optional (default=1e6)
+        Total umber of documents. It is only used in `partial_fit` method.
```
Total number of documents. It is only used in the ``partial_fit`` method.

amueller replied:
Why is this needed? It seems a bit odd.

amueller replied:
Also, the default value seems very arbitrary...

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+        Total umber of documents. It is only used in `partial_fit` method.
+
+    batch_size : int, optional (default=128)
+        Number of documents to use in each EM iteration. Only used in online learning method.
```
Only used in online learning.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+
+    evaluate_every : int optional (default=-1)
+        How often to evaluate perplexity. Only used in `fit` method.
+        set it to -1 to not evalute perplexity in training at all.
```
Any reason to use -1 and not 0?

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+        will also increase total training time.
+        Evaluating perplexity in every iteration will increase training time up to 2X.
+
+    normalize_doc : boolean, optional (default=False)
```
Is this described in the paper? I don't remember seeing that....

amueller replied:
Never mind, I thought this was used in fitting, but it is only a post-processing step.

chyikwei replied:
Yeah. `normalize_doc ` is not used in the algorithm, and user can do this easily.
going to remove it.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+        If `True`, sum of topic distribution will be 1.0 for each document.
+
+    perp_tol : float, optional (default=1e-1)
+        Perplexity tolerance in batch learning.
```
What happens if evaluate_every=-1? Is it just ignored? If so, this should be documented.

amueller commented:
All occurences of "Bayes" or "Bayesian" should probably be capitalized.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+    ----------
+    components_ : array, [n_topics, n_features]
+        Topic word distribution. components_[i, j] represents word `j` in topic `i`.
+        In the literature, this is called `lambda`, and we can calcuate `E[log(beta)]` from it.
```
I'm not sure what the second part of the sentence means here.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+                 max_doc_update_iter=100, n_jobs=1, verbose=0,
+                 random_state=None):
+        self.n_topics = n_topics
+        if n_topics <= 0:
```
please do all input validation in the ``fit`` method.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+                learning_method)
+
+        self.doc_topic_prior = doc_topic_prior
+        if self.doc_topic_prior is None:
```
please assign all variables like these in the ``fit`` method.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+            self.topic_word_prior_ = self.topic_word_prior
+
+        self.learning_method = learning_method
+        if learning_method not in ('batch', 'online'):
```
again, input validation should go into the fit method.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+        """
+        Initialize latent variables.
+        """
+        self.rng = check_random_state(self.random_state)
```
rng should probably have a trailing underscore.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+            Document word matrix.
+
+        cal_diff : boolean
+            parameter that indicate to calculate differene in `component_` or not.
```
Whether to calculate the change in ``components_`` or not.

amueller replied:
maybe a more natural shorthand would be ``calc_diff`` or ``calc_change`` here and also in the ``_update_doc_distribution`` function.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+
+        """
+
+        # parell run e-step
```
Run e-step in parallel

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+    def _em_step(self, X, total_samples, batch_update):
+        """
+        EM update for 1 iteration
+        update `_component` by bath VB or online VB
```
batch

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+        EM update for 1 iteration
+        update `_component` by bath VB or online VB
+
+        parameters
```
capitalize

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+        """
+        E-step
+
+        parameters
```
capitalize

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+            Document word matrix.
+
+        total_samples : integer
+            Total umber of document. It is only used when
```
Total number of documents

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+            batch_update is `False`.
+
+        batch_update : boolean
+            parameter that control updating method.
```
controls

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+        Returns
+        -------
+        doc_topic_distr : array, shape=(n_samples, n_topics)
+            Unnormailzed document topic distribution.
```
Unnormalized.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+        """
+        X = check_array(X, accept_sparse='csr')
+        check_non_negative(X, whom)
+        if not sp.issparse(X):
```
Ah, so my comment about dense matrices was not entirely correct.... The sparse format is only used in a few places, right? Have you tried running on dense data?

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+
+    def fit_transform(self, X, y=None):
+        """
+        Learn a model for X and returns the transformed data
```
Either add an "s" to "Learn" or remove it from "returns".

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+
+        if n_features != self.n_features:
+            raise ValueError(
+                "feature dimension(vocabulary size) doesn't match.")
```
space after dimension.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+    def fit(self, X, y=None):
+        """
+        Learn model from X with variational bayes method.
+        when `learning_method` is 'online', use min batch udpate.
```
mini batch update.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+                doc_topics = self._em_step(X, total_samples=total_samples, batch_update=False)
+
+            # check perplexity
+            if evaluate_every > 0 and (i + 1) % evaluate_every == 0:
```
So if evaluate_every == -1, it will always run ``max_iter`` iterations?

larsmans replied:
Also this checks `evaluate_every > 0`, so actually 0 means the same as -1.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+            raise NotFittedError(
+                "no 'components_' attribute in model. Please fit model first.")
+
+        # make sure word size is the same in fitted model and new doc
```
word size? The number of features, right?

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+        """
+        Calculate approximate bound for data X and document topic distribution
+        Since log-likelihood cannot be computed directly, we use this bound
+        estimate it.
```
to estimate it.

amueller replied:
The original implementation says "Estimates the variational bound over *all documents* using only the documents passed in as "docs."" which sounds different from what it says here.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+    def perplexity(self, X, doc_topic_distr, sub_sampling=False):
+        """
+        calculate approximate perplexity for data X and topic distribution `gamma`.
+        Perplexity is defined as exp(-1. * log-likelihood per word)
```
According to [wikipedia](https://en.wikipedia.org/wiki/Perplexity) it is 2 ** entropy

larsmans replied:
@amueller WP uses a log2 version of entropy. We use natural logs everywhere.

amueller replied:
ok but entropy is not the same as log-likelihood per word, right?

chyikweiyau replied:
For perplexity, the definition and calculation is from the [literature](https://www.cs.princeton.edu/~blei/papers/HoffmanBleiBach2010b.pdf). (formula 15 & 16)

[here](http://stats.stackexchange.com/questions/18167/how-to-calculate-perplexity-of-a-holdout-with-latent-dirichlet-allocation) is some discussion on stackexchange .

amueller replied:
Hm ok I didn't read the full wiki article. Apparently there is a difference between the perplexity of a distribution and the perplexity of a model of a distribution (which from outside the community seems slightly confusing), and the perplexity of a model doesn't have the ``p(w)`` term... oh well..

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+
+        Parameters
+        ----------
+        X : sparse matrix, [n_samples, n_features]
```
Actually you accept dense matrices, but convert them to sparse.

amueller replied:
(same for the other public methods).

amueller commented on sklearn/decomposition/tests/test_online_lda.py:
```diff
@@ -0,0 +1,292 @@
+
+
+def test_lda_fit_batch():
+    """
```
can you please us comments instead of ``"""`` in all tests?

amueller commented on sklearn/decomposition/tests/test_online_lda.py:
```diff
@@ -0,0 +1,292 @@
+
+
+def _build_sparse_mtx():
+    # Create 3 topics and each have 3 words
```
maybe say "words belong only to a single topic" ?

amueller commented on sklearn/decomposition/tests/test_online_lda.py:
```diff
@@ -0,0 +1,292 @@
+                                    learning_method='batch', random_state=rng)
+    lda.fit(X)
+
+    correct_idx_grps = [(0, 1, 2), (3, 4, 5), (6, 7, 8)]
```
you could additionally check ``adjusted_rand_score([0, 0, 0, 1, 1, 1, 2, 2, 2], np.argmax(lda.transform(X), axis=1) == 1``

amueller commented on sklearn/decomposition/tests/test_online_lda.py:
```diff
@@ -0,0 +1,292 @@
+
+    correct_idx_grps = [(0, 1, 2), (3, 4, 5), (6, 7, 8)]
+    for c in lda.components_:
+        top_idx = set(c.argsort()[-3:][::-1])
```
This line is not very self-explanatory. Could you add a coment?

amueller commented on sklearn/decomposition/tests/test_online_lda.py:
```diff
@@ -0,0 +1,292 @@
+    """
+    rng = np.random.RandomState(0)
+    n_topics, doc_topic_prior, topic_word_prior, X = _build_sparse_mtx()
+    lda = LatentDirichletAllocation(n_topics=n_topics, doc_topic_prior=doc_topic_prior,
```
this uses the default number of total documents, right? is that on purpose?

chyikwei replied:
No. This is not on purpose. I will add `total_samples` parameter here.

amueller commented on sklearn/decomposition/tests/test_online_lda.py:
```diff
@@ -0,0 +1,292 @@
+    lda = LatentDirichletAllocation(n_topics=n_topics, doc_topic_prior=doc_topic_prior,
+                                    topic_word_prior=topic_word_prior, random_state=rng)
+    X_trans = lda.fit_transform(X)
+    assert_true((X_trans > 0.0).any())
```
Why don' you use the same input as for the other tests and call ``toarray()`` on it? Or are you testing something else here?

chyikwei replied:
Oh. The test name is wrong. I was trying to test the result of transform.
Will fix it and add a separate test for dense input.

amueller commented on sklearn/decomposition/tests/test_online_lda.py:
```diff
@@ -0,0 +1,292 @@
+    """
+    for method in ('online', 'batch'):
+        rng = np.random.RandomState(0)
+        n_topics, doc_topic_prior, topic_word_prior, X = _build_sparse_mtx()
```
I feel that this data is a bit too trivial for this check... I don't have a great alternative, though.

amueller commented on sklearn/decomposition/tests/test_online_lda.py:
```diff
@@ -0,0 +1,292 @@
+                                    topic_word_prior=topic_word_prior, learning_offset=5.,
+                                    total_samples=20, random_state=rng)
+    for X in [X_1, X_2]:
+        lda.partial_fit(X)
```
can you use ``assert_raises_regex`` on the second call to partial fit? That would be more explicit than the ``raises`` decorator.

amueller commented on sklearn/decomposition/tests/test_online_lda.py:
```diff
@@ -0,0 +1,292 @@
+    lda = LatentDirichletAllocation(learning_method='unknown')
+
+
+@raises(ValueError)
```
At least in the multi-line tests I would prefer ``assert_raises_regex`` to the decorator.

amueller commented on sklearn/decomposition/tests/test_online_lda.py:
```diff
@@ -0,0 +1,292 @@
+    assert_almost_equal(perplexity_1, perplexity_2)
+
+
+if __name__ == '__main__':
```
can you please remove this?

ogrisel replied:
+1, to run the tests on a specific file you can use nose directly:

```
nosetests -s -v sklearn/decomposition/tests/test_online_lda.py
```

and for a specific test:

```
nosetests -s -v sklearn/decomposition/tests/test_online_lda.py:test_lda_score_perplexity
```

amueller commented:
This looks pretty good. I haven't checked the math in-depth yet, but I'll do that soon. I think it would be good to add some tests on slightly less trivial data maybe.
Also, I'm not sure how much we gain from the cython code.
Lastly, it looks to me as if supporting dense data shouldn't be too much work. If that is the case, I'd prefer not to convert it to sparse data.

amueller commented on examples/decomposition/topics_extraction_with_online_lda.py:
```diff
@@ -0,0 +1,71 @@
+"""
```
can you please run pep8 on this file?

amueller commented:
Maybe an example of online vs batch lda would be nice?

amueller commented:
Looking at line_profiler, your cython seems to be in exactly the  right place.
It looks a bit like cythonizing the whole inner loop in ``_update_doc_distribution`` that goes over max_iter might be worth it. Not necessary for the PR, but have you looked into that?
3% of time in this function is spent in the ``for``, 4% in ``last_d = doc_topic_d`` (I don't understand) and another 3% in ``if meanchange < mean_change_tol``

chyikwei commented:
@amueller thanks for finding all those issues. I will fix them soon.

For the Cython code, it is based on my previous [profiling result] (https://gist.github.com/chyikwei/e92c01bdb04450aa5b94):
1) For `meanchange` line, the difference is [33.7%](https://gist.github.com/chyikwei/e92c01bdb04450aa5b94#file-01_original-L63) vs [5.5%](https://gist.github.com/chyikwei/e92c01bdb04450aa5b94#file-02_add-mean_change-to-cython-L63).
2) For `_dirichlet_expectation` line, the difference is [48.7%](https://gist.github.com/chyikwei/e92c01bdb04450aa5b94#file-02_add-mean_change-to-cython-L60) vs [37.4%](https://gist.github.com/chyikwei/e92c01bdb04450aa5b94#file-03_update-_dirichlet_expectation-L62).
(Since I changed some code after that, I will run `line_profiler` again.)

For "cythonizing whole inner loop", I will check how to do that after fix other issues.

larsmans commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+
+    normalize_doc : boolean, optional (default=False)
+        Whether to normalize the topic distribution for each document.
+        If `True`, sum of topic distribution will be 1.0 for each document.
```
I find it strange that unnormalized distributions are produced by default.

larsmans commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+
+    def partial_fit(self, X, y=None):
+        """
+        Online VB with Min-Batch update.
```
This can be made more descriptive. E.g.,

```
Update the LDA model based on new documents X (mini-batch learning).
```

larsmans commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+
+    def fit(self, X, y=None):
+        """
+        Learn model from X with variational bayes method.
```
Newline here.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+            # online update
+            # In the literature, the weight is `rho`
+            weight = np.power(self.learning_offset + self.n_iter_, -self.learning_decay)
+            doc_ratio = float(total_samples) / X.shape[0]
```
are the total_samples mentioned in the paper? I couldn't find it ....

chyikwei replied:
In the [paper](https://www.cs.princeton.edu/~blei/papers/HoffmanBleiBach2010b.pdf), It is "D" in equation (8).

amueller replied:
ok, thanks

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+        X_indptr = X.indptr
+
+        # E[log p(docs | theta, beta)]
+        for d in xrange(0, n_samples):
```
This is the same as in the e-step, only not using the cython methods, right? Can we refactor that?

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,668 @@
+                tmax = temp.max()
+                norm_phi[i] = np.log(np.sum(np.exp(temp - tmax))) + tmax
+            score += np.sum(cnts * norm_phi)
+
```
This clearly closely resembles the code in the original implementation, but I have a hard time following this from the paper alone.... Did anyone go through this?

amueller commented:
@larsmans would you be comfortable reviewing the math? The maths is basically the same as in Matt Hoffman's implementation, so I'm not sure into who much detail we should go. On the other hand, I don't want black magic code in sklearn.

amueller commented:
Or maybe @mblondel knows about this?

chyikwei commented:
To-Do List:
- [ ] rebase master
- [ ] fix typos and import document
- [ ] move input validation and variable assignment to the `fit` method.
- [ ] move `check_non_negative` to `sklearn.utils.validation` (same for NMF)
- [ ] native dense input support
- [ ] testing code improvement
- [ ] refactor `_approx_bound` method (remove duplicate code)
- [ ] pep8
- [ ] profiling cython code
- [ ] example of online vs batch LDA

amueller commented:
Thanks for collecting the todo list, that is always a good thing.
You merged master, not rebased.
I would suggest you go back to before the merge, squash the commits, and then rebase.
Something like
```
git checkout master 
git pull upstream master # make sure master is up-to-date.
git checkout onlineldavb #go to the branch
git reset --hard 100ba3bc9ddb62b929bcb00bd9ecfaf6454c2b75 #discard the last merge commit
git rebase -i 5df19b4025e73b59574c1d067baea2b9616de831  # start interactive rebasing for squashing
# change all the "pick" to "squash" except for the first commit.
# only a single commit left in the branch
git rebase master # rebase on master
```

chyikwei commented:
@amueller Thanks! Don't know this rebase trick before.

amueller commented:
I want to add a bit more on git tricks to the developer documentation, but it's not there yet :-/ hope it is not too much of a hassle.

chyikwei commented:
It does take me some time to figure out how to do this.
Since I merged `master` several times, I got some conflicts during rebase.
So in the end I have to hard reset to another commits, cherry pick others, then rebase.

chyikwei committed:
fix typos and imporve documentation

chyikwei committed:
fix typo & improve doc 2

chyikwei committed:
testing code improvement

chyikwei committed:
move input validation and variable assignment to fit method

chyikwei committed:
fix _approx_bound with subsampling

amueller commented:
Thanks for the rebase and sorry for the hassle.

chyikwei committed:
native dense input support

amueller commented:
That looks good. have you tried with any dense data?

chyikwei commented:
I only tried convert 20 news groups data to dense and run LDA on it.
haven't try other dataset yet.

amueller commented:
Ok. I think we shouldn't worry too much about making it faster right now, we can always add more cythonization later if it pays of. I'd rather merge this sooner so we can get some user feedback.

larsmans commented:
+1

I'm willing to try this out and optimize it where needed.

amueller commented:
Btw, I emailed with Matt and he has some weave code lying around against which we could compare. @larsmans : could you review the PR and see if you have other comments that are not addressed on @chyikwei's check-list above?

chyikwei commented:
ok. So should I remove `profiling cython code` from to-do list?
I can finish other tasks in the weekend.

amueller commented:
Yeah, I think it would be great to wrap this up. You can always profile more after it was merged if you are so inclined.

chyikwei committed:
rename variable and minor pep8 fix

chyikwei committed:
clean up test and add error check in preplexity method

chyikwei committed:
remove duplicate code in _approx_bound method

chyikwei committed:
move check_non_negative to utils/validation.py

chyikwei committed:
fix params and import test coverage

chyikwei committed:
update example and change total_samples in fit method

chyikwei commented:
All tasks in the to-do list are done now.

amueller commented:
I just realized there is no entry in the user guide. Would you be willing to write one?
It ``should be in doc/modules/decomposition.rst``, briefly explaining what the algorithm does and what important parameters are.

chyikwei commented:
sure. will do it soon.

amueller commented:
thanks :)

chyikwei committed:
add LDA user guide

chyikwei committed:
add LDA to classes.rst

chyikwei committed:
random initialize in e-step

chyikwei commented:
added LDA user guide

amueller commented:
Great, thank you so much. I'll try a last round of reviews soon, but I think this is good to go. I hope someone else might find the time for the second review.

GaelVaroquaux commented on doc/modules/decomposition.rst:
```diff
@@ -726,3 +726,53 @@ the data.
+
+Latent Dirichlet Allocation is a generative probabilistic model for collections of discrete
+dataset such as text corpora. It is also a topic model that used for discovering abstract
+topics from a collection of docuements.
```
Typo: documents

GaelVaroquaux commented on doc/modules/decomposition.rst:
```diff
@@ -726,3 +726,53 @@ the data.
+Unlike batch method, online method computes intermediate variational variables from mini-batch
+data points and updates the variational variables as a weighted average of the intermediate
+variables and their current values.
+Therefore, online method usually converges faster than batch method.
```
I think that you are going too much into details here: what the end user really cares about is that it is fast, but not really about the details of why.

GaelVaroquaux commented on examples/decomposition/topics_extraction_with_online_lda.py:
```diff
@@ -0,0 +1,97 @@
+    print()
+
+
+def lda_example():
```
I'd rather not have a function + call in __main__ but as a script, as in the other examples.

GaelVaroquaux commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,699 @@
+    [1] "Online Learning for Latent Dirichlet Allocation", Matthew D. Hoffman,
+        David M. Blei, Francis Bach, 2010
+
+    [2] "Stochastic Variational Inference", Matthew D. Hoffman,, David M. Blei,
```
Typo: two comas

GaelVaroquaux commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,699 @@
+
+        """
+
+        # Run e-step in parallel
```
This is buggy: it will not implement joblib semantics for n_jobs=-2. Please use sklearn.externals.joblib.cpu_count for that.

larsmans replied:
@GaelVaroquaux Is this comment still current?

GaelVaroquaux commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,699 @@
+        """
+        self._check_params()
+        X = self._check_non_neg_array(X, "LatentDirichletAllocation.fit_transform")
+        return self.fit(X).transform(X)
```
I don't see the necessity of defining this method: I don't think that it provides much value compared to the inherited fit_transform method from the TransformerMixin.

chyikwei committed:
change LDA example to script

chyikwei committed:
change doc, fix n_jobs, and remove fit_transform

chyikwei commented:
@GaelVaroquaux thanks for the comments.

The following issues are fixed:
1. change example to script.
2. remove `fit_transform` method.
3. use `utils._get_n_jobs` to get `n_jobs` value.
4. update user guide and fix document typo.

ogrisel commented on doc/modules/decomposition.rst:
```diff
@@ -726,3 +726,51 @@ the data.
+.. topic:: Examples:
+
+    * :ref:`example_decomposition_topics_extraction_with_online_lda.py`
+
```
Could you please add a mathematical formulation section that gives a high level formal description of the probabilistic model being estimated by this class? You don't need to give the detailed derivation of the estimation procedure but just a description of the generative model (e.g. a plate diagram along with the formula of the PDF of the distributions used in the diagram).

ogrisel commented:
I think it should be possible to release the CIL in the Cython routines. Could you please give it a try?

ogrisel commented:
Also I think we should merge the existing NMF example on topic modeling with the new LDA example contributed in this PR. There are giving similar results on 20 newsgroups: it's good to have examples that compares methods on similar problems. Many people know that LDA can be used for topic modeling but far fewer know that NMF is actually quite useful too in that regard.

ogrisel commented:
Actually it might be possible to write most of the `_update_doc_distribution` in Cython under a `with nogil` block which would allow to use the `backend='threading'` mode of joblib which would remove useless memory copy or memory mapping and also skip the multiprocessing communication overhead. But I would not put that re-factoring as a requirements to accept this PR, in can be done later.

ogrisel commented:
@chyikwei could you please run this code on a bigger dataset, like a large subset of wikipedia and compare the perplexity and training time of your model with the the original `onlineldavb` implementation, the alternative LDA implementation by @ariddell and maybe also the one in gensim configured with the same number of topics and similar values for `max_iter`?

amueller commented:
Wouldn't the code by @ariddell be to slow for a large dataset?
The results all looked good and I think we should merge this asap, and polish later. The results comparing to gensim look very good, so I'm not to worried.

amueller commented:
Having a comparison of NMF and LDA would be cool, though.

ogrisel commented:
Alright, so new priorities:

- merge the example with the NMF example
- add a mathematical formulation to the doc
- do a final review on the code organization: I am pretty sure we can get rid of the `.rng_` attribute and follow the usual pattern: `random_state = check_random_state(self.random_state)` in `fit`  and then pass the local variable `random_state` as argument to the private methods. I would need to double check the code though.

amueller commented:
awsome :)

chyikwei commented:
ok. I will update doc and merge NMF example.
Where should I put the merged example? `examples/applications/` or `examples/decomposition/`?

amueller commented:
I'd say leave it in applications.

chyikwei commented:
ok. thanks!

chyikwei committed:
merge NMF and LDA example

ogrisel commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,678 @@
+
+    def partial_fit(self, X, y=None):
+        """
+        Online VB with Min-Batch update.
```
Typo: Mini-Batch

Also please put the first line on the first of the docstring to follow PEP257:

```python
def partial_fit(self, X, y=None):
    """Online VB with Mini-Batch update.
    
    ... remaining paragraphs for the docstring
    """

This is useful to get IDE and the sphinx documentation properly pick up the first line as a short descriptions / tool tip for the method.
```

ogrisel commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,678 @@
+        """
+        self.rng_ = check_random_state(self.random_state)
+        self.n_iter_ = 1
+        self.n_features = n_features
```
This should be `self.n_features_` as it is set from the data and not as an hyper-parameter in the constructor.

ogrisel replied:
But actually there is no need to store this, we can always access `self.components_.shape[1]` if we need later instead.

ogrisel commented:
> do a final review on the code organization: I am pretty sure we can get rid of the .rng_ attribute and follow the usual pattern: random_state = check_random_state(self.random_state) in fit and then pass the local variable random_state as argument to the private methods. I would need to double check the code though.

Ok I re-read the code, and indeed at least when we are using the `partial_fit` method we need to maintain some PRNG state for the E-step across successive `partial_fit` calls. So I see no alternative than to store the rng instance as an attribute of the estimator. However I would find it more consistent with the library to rename `self.rng_` to `self.random_state_`.

chyikwei committed:
improve LDA doc

chyikwei committed:
fix typo and improve doc

chyikwei committed:
use n_features_

chyikwei commented:
I changed the followings:
* merge the example with the NMF example
* add a mathematical formulation to the doc
* fix typo and rename a few parameters.

Let me know if I missed something.

ogrisel commented:
It seems that some comments by @amueller were not yet addressed, see the diff for more details: https://github.com/scikit-learn/scikit-learn/pull/3659/files

ogrisel commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,671 @@
+            self._init_latent_vars(n_features)
+
+        if n_features != self.n_features_:
+            raise ValueError("Feature dimension (vocabulary size) doesn't match.")
```
Please make this message more explicit by including expected and observed values:

```python
raise ValueError("The provided data has %d dimensions while the"
    " model was trained on data with vocabulary size %d."
    % (n_features, self.components_.shape[1]))
```

ogrisel commented:
Please also rename `self.rng_` to `self.random_state_` to be consistent with other probabilistic models.

ogrisel commented:
Also I think we should make sure that the `score` method is in line with the ongoing work by @xuewei4d on (Bayesian) GMMs. It would be grat @xuewei4d if you could do a review of this PR to check what could be possible source inconsistencies between your GSoC models and this one.

chyikwei committed:
remove self.n_features_

chyikwei committed:
remove main in test

chyikwei committed:
make dimension mismatch err msg more explicit

chyikwei committed:
change all rng to random_state

xuewei4d commented:
I did a quick review on the ```score``` method. Because I didn't consider ```BayesianGaussianMixture``` when coding ```DensityMixin``` and ```GaussianMixture```, current ```DensityMixin``` is not compatible with any variational models. It has to compute additional term to the score, i.e. the score comes from variational distributions. Anyway, the lower bound or approximation bound is the right thing we need to compute for variational methods.

ogrisel commented:
@chyikwei could you please answer or address the comments of @amueller on the `_online_lda.pyx` file: https://github.com/scikit-learn/scikit-learn/pull/3659/files ?

chyikwei commented:
sure. I will benchmark the cython code again since there are some code changes after I run `line_profiler` last time.

chyikwei commented:
Here is the cython code benchmark (on `_update_doc_distribution` function)

1. `mean_change` improvement: from [32.8%] (https://gist.github.com/chyikwei/2b2c68b9f9c46e4866c6#file-without-cython-code-L57) to [6%](https://gist.github.com/chyikwei/2b2c68b9f9c46e4866c6#file-with-cython-code-L56). ([complete result](https://gist.github.com/chyikwei/2b2c68b9f9c46e4866c6))

2. `_log_dirichlet_expectation` improvement: from [38.4%](https://gist.github.com/chyikwei/91c374cf665b5c70482b#file-without-cython-code-L55) to [28.9%](https://gist.github.com/chyikwei/91c374cf665b5c70482b#file-with-cython-code-L53). ([complete result](https://gist.github.com/chyikwei/91c374cf665b5c70482b))

For reference, [here](https://gist.github.com/chyikwei/91c374cf665b5c70482b#file-ipython_profiling_script-py) is my profiling code.

chyikwei commented:
And should I move cython decorators to top of the file?

I see some files use `#cython: boundscheck=False`([example](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/_logistic_sigmoid.pyx#L1-L3)) and some use `cython.boundscheck(False)`([example](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/sparsefuncs_fast.pyx#L20-L22)).

chyikwei committed:
fix feature_names

chyikwei committed:
improve document in _online_lda.pyx

xuewei4d commented on sklearn/decomposition/online_lda.py:
```diff
@@ -0,0 +1,676 @@
+                cnts = X[idx_d, ids]
+            temp = dirichlet_doc_topic[idx_d, :, np.newaxis] + self.dirichlet_component_[:, ids]
+            tmax = temp.max(axis=0)
+            norm_phi = np.log(np.sum(np.exp(temp - tmax), axis=0)) + tmax
```
scikit-learn has a logsumexp function in utils.extmath

chyikwei replied:
cool. will use it. thanks!

chyikwei committed:
use logsumexp

amueller commented:
I haven't checked the documentation in detail but I'd be happy to merge this now. @ogrisel what do you think?

amueller commented:
ping @ogrisel again ;)

amueller commented:
@larsmans are you interested in this?

larsmans commented:
Merged this. Let's finish any nitpicking in master.

amueller commented:
Thanks @larsmans, I agree :) :beers: Thank you so much @ chyikwei !

mblondel commented:
It has been suggested to mention more clearly in the code that Matt Hoffmann allowed us to license the code as BSD even though it's derived from his GPL implementation:
https://twitter.com/EdwardRaffML/status/631123381212082176

larsmans commented:
Yeah, it'd be good to reproduce the email or something. @chyikweiyau, could you do that?

chyikwei commented:
@larsmans sure. [Here](https://gist.github.com/chyikwei/14778a742b2b8d60dd7d) is the email I sent to Matt Hoffmann before.
And he emailed me relicensed code. [Here](https://www.dropbox.com/s/flv6yyk67ywejch/onlineldavb.tar?dl=0) are the files.

amueller commented:
Maybe just add to the licence header "relicenced as BSD with the kind permission of Matt Hoffmann"?
